gammapy/spectrum/energy_group.py

```diff
@@ -0,0 +1,534 @@
+        return len(self.groups)
+
+    # Methods to compute total ranges
+
+    def compute_range_safe(self):
+        """Apply safe energy range to ``groups``.
+        """
+        emin = self.obs.lo_threshold
+        emax = self.obs.hi_threshold
+        self.set_energy_range(emin=emin, emax=emax)
+
+    def set_energy_range(self, emin=None, emax=None):
+        """Apply energy range to ``groups``.
+        """
+        if emin:
+            self.groups.apply_energy_min(emin)
+        if emax:
+            self.groups.apply_energy_max(emax)
+
+    # Methods to compute groupings
+
+    def compute_groups_fixed(self, ebounds):
+        """Compute grouping for a given fixed energy binning.
+        """
+        self.groups.apply_energy_min(energy=ebounds[0])
+        # print(self.groups)
+        # import IPython; IPython.embed()
+        self.groups.apply_energy_max(energy=ebounds[-1])
+        self.groups.apply_energy_binning(ebounds=ebounds)
+
+    def compute_groups_npoints(self, npoints):
+        """TODO: document"""
+        emin, emax = self.get_safe_range()
+        npoints = self.config['n_points']
+        return EnergyBounds.equal_log_spacing(
+            emin=emin, emax=emax, nbins=npoints,
+        )
+
+
+class SpectrumEnergyGroup(object):
+    """Spectrum energy group.
+
+    Represents a consecutive range of bin indices (both ends inclusive).
+    """
+    valid_bin_types = ['normal', 'underflow', 'overflow']
+    """Valid values for ``bin_types`` attribute."""
+
+    def __init__(self, energy_group_idx, bin_idx_min, bin_idx_max, bin_type,
+                 energy_min, energy_max):
+        self.energy_group_idx = energy_group_idx
+        self.bin_idx_min = bin_idx_min
+        self.bin_idx_max = bin_idx_max
+        if bin_type not in self.valid_bin_types:
+            raise ValueError('Invalid bin type: {}'.format(bin_type))
+        self.bin_type = bin_type
+        self.energy_min = energy_min
+        self.energy_max = energy_max
+
+    def to_dict(self):
+        """Data as `~collections.OrderedDict`."""
+        data = OrderedDict()
+        data['energy_group_idx'] = self.energy_group_idx
+        data['bin_idx_min'] = self.bin_idx_min
+        data['bin_idx_max'] = self.bin_idx_max
+        data['bin_type'] = self.bin_type
+        data['energy_min'] = self.energy_min
+        data['energy_max'] = self.energy_max
+        return data
+
+    @property
+    def bin_idx_range(self):
+        """Range of bin indices (both sides inclusive)."""
+        return self.bin_idx_min, self.bin_idx_max
+
+    @property
+    def energy_range(self):
+        """Energy range."""
+        return EnergyRange(min=self.energy_min, max=self.energy_max)
+
+    @property
+    def bin_idx_list(self):
+        """List of bin indices in the group."""
+        left, right = self.bin_idx_range
+        return list(range(left, right + 1))
+
+    def __str__(self):
+        return str(self.to_dict())
+
+
+class SpectrumEnergyGroups(list):
```

Comment:
It's hard to see the purpose of this class without much documentation. At first sight it looks super-complicated. In the end we just want a list with integers to put each fine energy bin into one flux points group, this list then goes to the flux point computation method. Can you add some explanation what this is for?